File: samples/Kawa.Sample.Mixed.UseCases/CreateUserUseCase.fs
namespace Kawa.Sample.Mixed.UseCases

open System
open System.Threading
open System.Threading.Tasks
open Kawa.Abstractions
open Kawa.FSharp

/// <summary>Creates a simple user response from the mixed sample F# use case.</summary>
type CreateUserUseCase() =
    interface IUseCase<CreateUserRequest, CreateUserResponse> with
        /// <inheritdoc />
        member _.ExecuteAsync(request: CreateUserRequest, cancellationToken: CancellationToken) =
            cancellationToken.ThrowIfCancellationRequested()

            CreateUserResponse(Guid.NewGuid().ToString("N"), request.Name)
            |> KawaResult.success
            |> Task.FromResult

File: samples/Kawa.Sample.CSharp/CreateUserUseCase.cs
using Kawa.Abstractions;

/// <summary>
/// Creates a simple sample user response.
/// </summary>
public sealed class CreateUserUseCase : IUseCase<CreateUserRequest, CreateUserResponse>
{
    /// <inheritdoc />
    public Task<KawaResult<CreateUserResponse>> ExecuteAsync(
        CreateUserRequest request,
        CancellationToken cancellationToken = default)
    {
        var response = new CreateUserResponse($"Created user {request.Name}.");
        return Task.FromResult(KawaResult<CreateUserResponse>.Success(response));
    }
}
